src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+    for(size_t col = 0; col != dst.n_cols; ++col)
+    {
+      size_t x = static_cast<size_t>(col * x_ratio);
+      double const x_diff = (col * x_ratio) - x; //distance of the nearet pixel(x axis)
+      double const x_diff_2 = 1 - x_diff;
+      double const y2_cross_x2 = y_diff_2 * x_diff_2;
+      double const y2_cross_x = y_diff_2 * x_diff;
+      double const y_cross_x2 = y_diff * x_diff_2;
+      double const y_cross_x = y_diff * x_diff;
+      dst(row, col) = y2_cross_x2 * src(y, x) +
+              y2_cross_x * src(y, x + 1) +
+              y_cross_x2 * src(y + 1, x) +
+              y_cross_x * src(y + 1, x + 1);
+    }
+  }
+
+  return dst;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+sepFilter2D(CubeType &InOutImage, arma::vec &kernel, size_t radius)
+{
+  CubeType OutImage = this->CopyMakeBorder(InOutImage, radius, radius, radius, radius);
+
+  arma::vec row_res, col_res;
+  // reverse InOutImage and OutImage to avoid making an extra matrix.
+  // InImage is renamed to InOutImage in this function for this reason only.  
+  arma::mat k_mat = kernel * kernel.t();
+  for(size_t k = 0; k < OutImage.n_slices; ++k)
+  {
+    for(size_t j = radius; j < OutImage.n_cols - radius; ++j)
+    {
+      for(size_t i = radius; i < OutImage.n_rows - radius; ++i)
+      {
+        InOutImage(i - radius, j - radius, k) = 
+            arma::accu(OutImage.slice(k).submat(i - radius, j - radius, i + radius, j + radius) % k_mat);
+      }
+    }
+  }
+
+  return InOutImage;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+ConvTriangle(CubeType &InImage, size_t radius)
+{
+  if (radius == 0)
+  {
+    return InImage;
+  }
+  else if (radius <= 1)
+  {
+    const double p = 12.0 / radius / (radius + 2) - 2;
+    arma::vec kernel = {1 , p, 1};
+    kernel /= (p + 2);
+    
+    return this->sepFilter2D(InImage, kernel, radius);
+  }
+  else
+  {
+    const size_t len = 2 * radius + 1;
+    arma::vec kernel(len);
+    for( size_t i = 0; i < radius; ++i)
+      kernel(i) = i + 1;
+    
+    kernel(radius) = radius + 1;
+    
+    size_t r = radius;
+    for( size_t i = radius + 1; i < len; ++i)
+      kernel(i) = r--;
+
+    kernel /= std::pow(radius + 1, 2);
+    return this->sepFilter2D(InImage, kernel, radius);
+  }
+}
+
+//just a helper function, can't use it for anything else
+//finds max numbers on cube axis and returns max values,
+// also stores the locations of max values in Location
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+MaxAndLoc(CubeType &mag, arma::umat &Location) const
+{
+  /*Vectorize this function after prototype works*/
+  MatType MaxVal(Location.n_rows, Location.n_cols);
+  for(size_t i = 0; i < mag.n_rows; ++i)
+  {
+    for(size_t j = 0; j < mag.n_cols; ++j)
+    {
+      /*can use -infinity here*/
+      double max =  std::numeric_limits<double>::min();
```

Comment:
``std::numeric_limits<double>::min()`` is ``DBL_MIN``: ``DBL_MIN`` returns the minimum positive normalized value, I guess what you like is: ``-DBL_MAX``?